stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1869,189 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
 }
+
+// SWIFT_ENABLE_TENSORFLOW
+extension Array where Element : Differentiable {
+  /// Views the array as the differentiable product manifold of `Element` multiplied with itself
+  /// `count` times.
+  @_fixed_layout
+  public struct DifferentiableView : Differentiable {
+    private var base: [Element]
+
+    /// The viewed array.
+    // I'm implementing this as a computed property instead of directly exposing `base` because the
+    // `@differentiable` annotation does not make the stored property actually differentiable. I
+    // think this is a bug. Maybe it's related to `@_fixed_layout`?
+    // TODO: Determine if that is a bug, and fix.
+    public var array: [Element] {
+      @differentiable(wrt: self, vjp: _vjpArray)
+      get {
+        return base
+      }
+      set(newValue) {
+        base = newValue
+      }
+    }
+
+    public func _vjpArray() -> ([Element], (Array<Element>.CotangentVector) -> CotangentVector) {
+      return (array, { $0 })
+    }
+
+    /// Construct a view of the given array.
+    @differentiable(wrt: array, vjp: _vjpInit)
+    public init(_ array: [Element]) { self.base = array }
+
+    public static func _vjpInit(_ array: [Element]) ->
+      (Array.DifferentiableView, (CotangentVector) -> CotangentVector)
+    {
+      return (Array.DifferentiableView(array), { $0 })
+    }
+
+    // MARK: - Differentiable conformance.
+
+    public typealias TangentVector = Array<Element.TangentVector>.DifferentiableView
+    public typealias CotangentVector = Array<Element.CotangentVector>.DifferentiableView
+    public typealias AllDifferentiableVariables =
+      Array<Element.AllDifferentiableVariables>.DifferentiableView
+
+    public var allDifferentiableVariables: AllDifferentiableVariables {
+      get {
+        return AllDifferentiableVariables(array.map { $0.allDifferentiableVariables })
+      }
+      set(newValue) {
+        precondition(
+          array.count == newValue.array.count,
+          "cannot set Array.DifferentiableView.AllDifferentiableVariables with count " +
+            "\(array.count) to Array.DifferentiableView.AllDifferentiableVariables with " +
+            "different count \(newValue.array.count)")
+        for i in array.indices {
+          array[i].allDifferentiableVariables = newValue.array[i]
+        }
+      }
+    }
+
+    public func moved(along direction: TangentVector) -> DifferentiableView {
+      precondition(
+        array.count == direction.array.count,
+        "cannot move Array.DifferentiableView with count \(array.count) along direction with " +
+          "different count \(direction.array.count)")
+      return DifferentiableView(zip(array, direction.array).map { $0.moved(along: $1) })
+    }
+
+    public func tangentVector(from cotangentVector: CotangentVector) -> TangentVector {
+      precondition(
+        array.count == cotangentVector.array.count,
+        "cannot use Array.DifferentiableView with count \(array.count) to get tangentVector from " +
```

Comment:
Are these over 80 columns?

Reply:
oops, I thought the limit was 100 based on some existing lines that go over 80

Reply:
Yes sorry that was my mistake. I assumed it was 100 because that's in the Google Swift Style Guide.